core/domain/stats_services.py

```diff
@@ -142,6 +142,11 @@ def _update_stats_transactional(
 
     for state_name, stats in aggregated_stats['state_stats_mapping'].items():
         if state_name not in exp_stats.state_stats_mapping:
+            # Some events in the past seems to have 'undefined' state names
+            # passed from the frontend code. These are invalid and should be
+            # discarded.
+            if state_name == 'undefined':
```

Comment:
I don't think any useful info can be logged here. I think the logging should happen at the controller. That will be done in a separate PR.

Reply:
Hm I think you should do it in this PR, because you're introducing the 'undefined' thing here. With this control flow we will completely lose all info about such errors.